[최지수] Step4 리뷰 요청합니다!

jisu1211 opened the pull request:
안녕하세요~ 근환님!

인수 테스트  기반 TDD 마지막 단계까지 잘 수행할 수 있도록 도와주셔서 감사합니다 ㅎㅎ
덕분에 코드가 많이 깔끔해지고 그리고 애매하게 알고있던 내용들까지도 확실하게 알고 가는 것 같습니다!

마지막 단계도 잘 부탁드립니다 ㅎㅎ 
감사합니다 😊

jisu1211 committed:
refactor: 사용하지 않는 import문 제거

jisu1211 committed:
refactor: 즐겨찾기 삭제시 회원 email 비교를 id 비교로 변경

jisu1211 committed:
docs: README.md 추가

jisu1211 committed:
refactor: distance type을 int로 변경

jisu1211 committed:
feat: 최단 거리에 따라 운임 비용을 계산하는 기능 추가
- 거리에 따라 기본거리/중거리/장거리를 구분하는 PathFare 추가
- 기본거리/중거리/단거리에 따라 운임 비용을 계산하는 PathFareCalculator 추가

test: PathFareCalculatorTest 추가, PathFareTest 추가

jisu1211 committed:
feat: Line 클래스에 Fare 추가
- 생성자 변경

test: 기존 Line 생성자 사용하던 부분 수정

jisu1211 committed:
refactor: 로그인 회원/익명 회원 추가
- AuthMember로 인증받도록 수정

jisu1211 committed:
feat: 지하철역을 포함하고 있는 구간의 호선을 구하는 로직 추가, 지하철 호선의 운임 비용 중 가장 큰 값을 구하는 로직 추가
test: LinesTest 추가, SectionsTest에 테스트 코드 추가

jisu1211 committed:
feat: 회원별 할인 정책 기능 구현
test: MemberDiscountPolicyTest 추가

jisu1211 committed:
feat: 지하철 운임 비용 계산 기능 구현
test: PathTest 추가, FareCalculatorTest 추가, PathServiceTest 수정

jisu1211 committed:
test: 지하철 운임 비용 계산 관련 인수 테스트 추가

jisu1211 committed:
docs: README.md 수정

geunhwanlee commented on src/main/java/nextstep/subway/auth/domain/AnonymousMember.java:
```diff
@@ -0,0 +1,18 @@
+    public Integer getAge() {
+        return 0;
+    }
+}
```
토큰이 없을때 `AnonymousMember`를 주는건 좋습니다 👍🏻
다만 `Controller`에서 보면
`AuthMember`가 `AnonymousMember`인지 `LoginMember`인지 알수가 없을것 같아요.

비즈지스 로직에서 `LoginMember`인지 확인하고 나이에 따른 할인정책을 적용하는게 더 명확하지 않을까 생각합니다 😀

jisu1211 replied:
넵 좀 더 고민해보도록 하겠습니다 ㅎㅎ

geunhwanlee commented on src/main/java/nextstep/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -26,6 +28,10 @@ public boolean supportsParameter(MethodParameter parameter) {
+        if(StringUtils.isEmpty(credentials)) {
+            return new AnonymousMember();
+        }
+
```
토큰이 없다면 예외가 발생해야하는 케이스도 있지 않을까요? 🤔

geunhwanlee commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -67,9 +67,9 @@ public LineResponse findLineResponseById(Long id) {
+    public void updateLine(Long id, LineRequest lineRequest) {
         Line persistLine = findLineById(id);
-        persistLine.update(new Line(lineUpdateRequest.getName(), lineUpdateRequest.getColor()));
+        persistLine.update(new Line(lineRequest.getName(), lineRequest.getColor(), lineRequest.getFare()));
```
`Line.update` 메서드에서 `fare`는 `update` 하지 않고있네요.
확인 부탁드려요 🙏

jisu1211 replied:
앗 수정하면서 놓쳤나보네요 ㅠ 꼼꼼하게 봐주셔서 감사합니다 ㅎㅎ

geunhwanlee commented on src/main/java/nextstep/subway/member/domain/MemberDiscountPolicy.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.util.function.Function;
+
+public enum MemberDiscountPolicy {
+    CHILD(6, 12, (totalFare) -> (int) Math.ceil((totalFare - 350) * 0.5)),
+    TEENAGER(13, 18, (totalFare) -> (int) Math.ceil((totalFare - 350) * 0.8)),
+    ADULT(19, Integer.MAX_VALUE, (totalFare) -> totalFare);
```
나이별 할인정책을 enum으로 표현하셨네요 👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/path/domain/Path.java:
```diff
@@ -7,18 +7,30 @@
+        this.fare = pathFare.getFare(this.distance);
+    }
+
+    public void calculateFare(int age, int lineFare) {
+        fare = FareCalculator.calculate(age, fare, lineFare);
```
1. 노선에 따른 가격정책 반영
2. 나이에 따른 가격정책 반영

이 순서대로 적용되는게 코드에 드러나면 좋을것 같아요! 🤔
```suggestion
    public void calculateFare(MemberDiscountPolicy memberDiscountPolicy, LineFarePolicy lineFarePolicy) {
        fare = lineFarePolicy.getFare(fare);
        fare = memberDiscountPolicy.getFare(fare);

```
이런 흐름은 어떨까요? 😀

jisu1211 replied:
넵 코드에서 반영되는 흐름이 잘 보이도록 수정해보겠습니다!

geunhwanlee commented on src/main/java/nextstep/subway/path/domain/PathFare.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.util.function.Function;
+
+public enum PathFare {
+    DEFAULT(10, (distance) -> PathFareCalculator.calculateDefaultFare()),
+    MIDDLE(50, (distance) -> PathFareCalculator.calculateMiddleFare(distance)),
+    LONG(Integer.MAX_VALUE, (distance) -> PathFareCalculator.calculateLongFare(distance));
```
거리별 가격 정책을 enum으로 표현하신것 좋네요 👍🏻

geunhwanlee reviewed (changes requested):
4단계 요금 조회 잘 해주셨습니다 👍🏻

다만 몇가지 고민할 부분이 있어서 코멘트 남겼어요.

확인해보시고 재 리뷰요청 부탁드려요 🙏

jisu1211 committed:
refactor: AuthMember에서 LoginMember, AnonymousMember 구분할 수 있도록 추가

jisu1211 committed:
fix: AuthenticationPrincipal에 required 옵션 추가, required가 false이고 인증 정보가 없으면 AnonymousMember 반환하도록 수정

jisu1211 committed:
fix: 로그인된 회원이 필요한 곳에서는 LoginMember를 사용하도록 수정

jisu1211 committed:
refactor: 운임 비용 계산 로직 리팩터링
- 로그인 회원인 경우에만 할인을 적용하도록 수정
- 운임 비용 계산시 호선 비용 반영, 할인 반영 로직이 명확하게 보이도록 수정
- 사용하지 않게된 FareCalculator 삭제
- 비회원일 경우에도 비용 계산을 할 수 있도록 @AuthenticationPrincipal required = false 옵션 적용

test: 로직 변경에 따른 테스트 코드 수정

jisu1211 committed:
fix: 호선 수정 시 비용 반영되지 않는 이슈 수정

jisu1211 committed:
test: 지하철 운임 비용 계산 테스트 케이스 추가

jisu1211 commented:
근환님 피드백 주신 부분들 고민해서 반영해봤습니다!

> 토큰이 없을때 AnonymousMember를 주는건 좋습니다 👍🏻
> 다만 Controller에서 보면
> AuthMember가 AnonymousMember인지 LoginMember인지 알수가 없을것 같아요.
> 
> 비즈지스 로직에서 LoginMember인지 확인하고 나이에 따른 할인정책을 적용하는게 더 명확하지 않을까 생각합니다 😀

AuthMember에서 `isLoginMember`를 통해서 로그인 회원인지 확인하고 MemberDiscountPolicy에서 로그인 회원이 아니면
할인을 적용하지 않는 로직으로 적용해봤습니다!

> 토큰이 없다면 예외가 발생해야하는 케이스도 있지 않을까요? 🤔

`@AuthenticationPrincipal`에 required 옵션을 추가해서 해당 옵션을 false로 사용할 경우에만 AuthMember를 사용하고 
인증이 항상 필요한 경우에는 LoginMember를 사용하도록 수정했습니다!

> 1. 노선에 따른 가격정책 반영
> 2. 나이에 따른 가격정책 반영

댓글에 달아주신 `LineFarePolicy`는 현재 로직 상에서 어떻게 구현해야 할지 애매해서 일단은 Path 안에 `Lines.findMaxFare()`를 옮겼는데
말씀해주신 의도와 같을지 잘 모르겠네요 😢

감사합니다 😊

geunhwanlee reviewed (approved):
리뷰 반영 잘 해주셨습니다 👍🏻

충분히 잘 해주셔서 머지하도록 할게요

인수테스트 기반 TDD 미션이 끝이 났네요 👏
쉽지 않았을텐데 고생 많으셨습니다.
그럼 남은 미션 마무리 잘 하세요 💪
